src/mscorlib/src/System/Threading/ExecutionContext.cs

```diff
@@ -191,18 +191,27 @@ internal static void SetLocalValue(IAsyncLocal local, object newValue, bool need
             if (previousValue == newValue)
                 return;
 
-            current = new ExecutionContext(current);
-            current.m_localValues[local] = newValue;
-
-            t_currentMaybeNull = current;
+            Dictionary<IAsyncLocal, object> newValues = new Dictionary<IAsyncLocal, object>(current.m_localValues);
+            newValues[local] = newValue;
 
+            List<IAsyncLocal> newChangeNotifications = current.m_localChangeNotifications;
             if (needChangeNotifications)
             {
                 if (hadPreviousValue)
-                    Contract.Assert(current.m_localChangeNotifications.Contains(local));
+                {
+                    Contract.Assert(newChangeNotifications.Contains(local));
+                }
                 else
-                    current.m_localChangeNotifications.Add(local);
+                {
+                    newChangeNotifications = new List<IAsyncLocal>(current.m_localChangeNotifications);
+                    newChangeNotifications.Add(local);
+                }
+            }
+
+            t_currentMaybeNull = new ExecutionContext(newValues, newChangeNotifications);
```

Comment:
Does a new ExecutionContext need to be created if its not null and not been captured yet? 

Or would something like this work:
```csharp
if (t_currentMaybeNull != null && !**captured_flag**)
{
    t_currentMaybeNull.m_localValues = newValues;
    t_currentMaybeNull.m_localChangeNotifications = newChangeNotifications;
}
else 
{
    t_currentMaybeNull = new ExecutionContext(newValues, newChangeNotifications);
}
```
Capture would have to set the flag.